runtime/parachains/src/disputes/slashing.rs

```diff
@@ -0,0 +1,757 @@
+	/// needed for slashing.
+	pub offenders: Vec<KeyOwnerIdentification>,
+	/// What fraction of the total exposure that should be slashed for
+	/// this offence.
+	pub slash_fraction: Perbill,
+	/// Whether the candidate was valid or invalid.
+	pub kind: SlashingOffenceKind,
+}
+
+impl<Offender> Offence<Offender> for SlashingOffence<Offender>
+where
+	Offender: Clone,
+{
+	const ID: Kind = *b"disputes:slashin";
+
+	type TimeSlot = DisputesTimeSlot;
+
+	fn offenders(&self) -> Vec<Offender> {
+		self.offenders.clone()
+	}
+
+	fn session_index(&self) -> SessionIndex {
+		self.time_slot.session_index
+	}
+
+	fn validator_set_count(&self) -> ValidatorSetCount {
+		self.validator_set_count
+	}
+
+	fn time_slot(&self) -> Self::TimeSlot {
+		self.time_slot.clone()
+	}
+
+	fn disable_strategy(&self) -> DisableStrategy {
+		match self.kind {
+			SlashingOffenceKind::ForInvalid => DisableStrategy::Always,
+			// in the future we might change it based on number of disputes initiated
+			SlashingOffenceKind::AgainstValid => DisableStrategy::Never,
+		}
+	}
+
+	fn slash_fraction(&self, _offenders: u32) -> Perbill {
+		self.slash_fraction
+	}
+}
+
+impl<KeyOwnerIdentification> SlashingOffence<KeyOwnerIdentification> {
+	fn new(
+		session_index: SessionIndex,
+		candidate_hash: CandidateHash,
+		validator_set_count: ValidatorSetCount,
+		offenders: Vec<KeyOwnerIdentification>,
+		kind: SlashingOffenceKind,
+	) -> Self {
+		let time_slot = DisputesTimeSlot::new(session_index, candidate_hash);
+		let slash_fraction = match kind {
+			SlashingOffenceKind::ForInvalid => SLASH_FOR_INVALID,
+			SlashingOffenceKind::AgainstValid => SLASH_AGAINST_VALID,
+		};
+		Self { time_slot, validator_set_count, offenders, slash_fraction, kind }
+	}
+}
+
+/// This type implements `SlashingHandler`.
+pub struct SlashValidatorsForDisputes<C> {
+	_phantom: sp_std::marker::PhantomData<C>,
+}
+
+impl<C> Default for SlashValidatorsForDisputes<C> {
+	fn default() -> Self {
+		Self { _phantom: Default::default() }
+	}
+}
+
+impl<T> SlashValidatorsForDisputes<Pallet<T>>
+where
+	T: Config<KeyOwnerIdentification = IdentificationTuple<T>>,
+{
+	/// If in the current session, returns the identified validators. `None`
+	/// otherwise.
+	fn maybe_identify_validators(
+		session_index: SessionIndex,
+		account_ids: &[AccountId<T>],
+		validators: impl IntoIterator<Item = ValidatorIndex>,
+	) -> Option<Vec<IdentificationTuple<T>>> {
+		// We use `ValidatorSet::session_index` and not
+		// `shared::Pallet<T>::session_index()` because at the first block of a new era,
+		// the `IdentificationOf` of a validator in the previous session might be
+		// missing, while `shared` pallet would return the same session index as being
+		// updated at the end of the block.
+		let current_session = T::ValidatorSet::session_index();
+		if session_index == current_session {
+			let fully_identified = validators
+				.into_iter()
+				.flat_map(|i| account_ids.get(i.0 as usize).cloned())
```

Comment:
We do shuffle validators in `SessionInfo` - is the `ValidatorIndex` really valid for indexing `account_ids`?

Reply:
Yes, account_keys are shuffled the same way:
https://github.com/paritytech/polkadot/blob/2eb7672905d99971fc11ad7ff4d57e68967401d2/runtime/parachains/src/session_info.rs#L166-L168